# R script - as input it takes a VCF file after having merged all individual VCF files generated by EH (merging_EH_VCF.py)

date ()
Sys.info ()[c("nodename", "user")]
commandArgs ()
rm (list = ls ())
R.version.string ## "R version 3.3.2 (2016-10-31)"

library(ggplot2); packageDescription ("ggplot2", fields = "Version") #"2.2.1"
library(grid); packageDescription ("grid", fields = "Version") #"3.3.2"
library(gridExtra); packageDescription ("gridExtra", fields = "Version") #"2.2.1"
library(reshape2); packageDescription ("reshape2", fields = "Version") #"1.4.2"
require(dplyr); packageDescription ("dplyr", fields = "Version") #"0.7.1"

# Function that takes the whole df (containing all the tsv files (all STR and pathogenic STR)) and takes the gene name as well as the start position
# Creates the STR distributions for each gene we pass to it, stratifying with the Illumina version V1 and V2 as well

# Function that prints into a file the raw data/numbers corresponding to the STR repeat-size frequencies plots
print_table_gene <- function(df_input, gene_name, output_folder){
  df_to_print = df_input %>% filter(gene %in% gene_name)
  df_to_print = df_to_print[,c(4,9)]
  colnames(df_to_print) = c('repeat-size', 'num_samples')
  output_table = paste(output_folder, paste(gene_name, 'freq_rawdata.tsv', sep = '_') , sep = '/')
  write.table(df_to_print, file = output_table, quote = F, row.names = F, col.names = T, sep = '\t')  
}


# Function that plots the STR repeat-size frequencies for a gene/locus across the cohort
plot_gene <- function(df_input, gene_name, gene_data_normal, gene_data_pathogenic, output_folder) {
  
  threshold_normal = gene_data_normal %>% filter(locus %in% gene_name) %>% select(threshold) %>% unlist() %>% unname()
  threshold_pathogenic = gene_data_pathogenic %>% filter(locus %in% gene_name) %>% select(threshold) %>% unlist() %>% unname()

  df_gene = df_input %>% filter(gene %in% gene_name)

	alt_number = df_gene$allele

	#df_gene_barplot = data.frame(number_repeats = alt_number, af = df_gene$AF)
	df_gene_barplot = data.frame(number_repeats = alt_number, af = df_gene$num_samples)

	# order by 'number of repetition'
	df_gene_barplot = unique(df_gene_barplot[order(df_gene_barplot[,1]),])

	rownames(df_gene_barplot) = df_gene_barplot$number_repeats

	df_gene_barplot$number_repeats = as.numeric(df_gene_barplot$number_repeats)
	
	pdf_name = paste(output_folder, gene_name, sep = "/")
	pdf_name = paste(pdf_name, 'pdf', sep = ".")

	min_value = min(df_gene_barplot$number_repeats)
	max_value = max(threshold_pathogenic + 1, df_gene_barplot$number_repeats)

	aux_plot = ggplot(unique(df_gene_barplot), aes(x = number_repeats, y = af)) + 
	  geom_bar(stat = "identity") + 
	  geom_text(aes(label=af), vjust=0, size = 4, colour = "grey") +
	  ylab("Allele frequency") + 
	  xlab("Repeat sizes (repeat units)") + 
	  ggtitle(gene_name) + 
	  geom_vline(xintercept = threshold_normal, colour = 'blue', lty = 2) + 
	  geom_vline(xintercept = threshold_pathogenic, colour = 'red', lty = 2) + 
	  coord_cartesian(xlim = c(min_value,max_value))
	  

	pdf(pdf_name)
	print(aux_plot)	
	dev.off()

}

# Represent in a plot, the allele frequency in Y-coor and the repeat-size in X-coor

# STR annotation, threshold including the largest normal and the smallest pathogenic sizes reported
gene_annotation_normal = './threshold_largest_normal_reported.txt'
gene_data_normal = read.table(gene_annotation_normal, stringsAsFactors=F, header = T)

gene_annotation_pathogenic = './threshold_smallest_pathogenic_reported.txt'
gene_data_pathogenic = read.table(gene_annotation_pathogenic, stringsAsFactors=F, header = T)


# Internal allele frequencies

# May 2018
# Read the VCF file, after merging all the individual VCF files derived by running the latest EH 
df = read.csv('/Users/kibanez/Documents/GEL_STR/STR_internal_allele_frequencies/EH-offtarget-v2.5.5-duplicatesRemoved/merged/merged_loci_internal_STR_frequencies_may2018_20632_V4.vcf', sep = '\t')

# August 2018
# Read the VCF file, after merging all the individual VCF files derived by running the latest EH 
df = read.csv('/Users/kibanez/Documents/GEL_STR/STR_internal_allele_frequencies/august2018/EH-offtarget-v2.5.5-duplicatesRemoved/merged/merged_loci_internal_STR_frequencies_august2018_30481_V4.tsv', sep = '\t')

# January 2019
# Read the VCF file (it's a TSV), after merging all the individual VCF files derived by running the latest EH 
df = read.csv('/Users/kibanez/Documents/GEL_STR/STR_internal_allele_frequencies/january2019/EH-offtarget-v2.5.5-duplicatesRemoved/merged/merged_loci_internal_STR_frequencies_may2018_42888_V4.tsv', sep = '\t')


#df = read.csv('/Users/kibanez/Documents/GEL_STR/GRCh38/17July2017/04082017_2200/merged_June2018/merged_july2017_june2018_5529_V4.vcf',sep = '\t')
                
# Folder where we want to save the plots - PDFs
output_folder = '/Users/kibanez/Documents/GEL_STR/STR_internal_allele_frequencies/january2019/EH-offtarget-v2.5.5-duplicatesRemoved/merged/plots'

dir.create(output_folder)

# The strategy we were following before was based on the start genomic coordinate, we now use the name of the gene.
plot_gene(df, 'C9orf72_GGGGCC', gene_data_normal, gene_data_pathogenic, output_folder)
plot_gene(df, 'AR_CAG', gene_data_normal, gene_data_pathogenic, output_folder)
plot_gene(df, 'ARX', gene_data_normal, gene_data_pathogenic, output_folder)
plot_gene(df, 'ATN1_CAG', gene_data_normal, gene_data_pathogenic, output_folder)
plot_gene(df, 'ATXN1_CAG', gene_data_normal, gene_data_pathogenic, output_folder)
plot_gene(df, 'ATXN2_CAG', gene_data_normal, gene_data_pathogenic, output_folder)
plot_gene(df, 'ATXN3_CAG', gene_data_normal, gene_data_pathogenic, output_folder)
plot_gene(df, 'ATXN7_CAG', gene_data_normal, gene_data_pathogenic, output_folder)
plot_gene(df, 'ATXN8OS', gene_data_normal, gene_data_pathogenic, output_folder)
plot_gene(df, 'ATXN10_ATTCT', gene_data_normal, gene_data_pathogenic, output_folder)
plot_gene(df, 'CACNA1A_CAG', gene_data_normal, gene_data_pathogenic, output_folder)
plot_gene(df, 'CNBP_CCTG', gene_data_normal, gene_data_pathogenic, output_folder)
plot_gene(df, 'CSTB_CCCCGCCCCGCG', gene_data_normal, gene_data_pathogenic, output_folder)
plot_gene(df, 'DAB1', gene_data_normal, gene_data_pathogenic, output_folder)
plot_gene(df, 'DIP2B', gene_data_normal, gene_data_pathogenic, output_folder)
plot_gene(df, 'DMPK_CTG', gene_data_normal, gene_data_pathogenic, output_folder)
plot_gene(df, 'FXN_GAA', gene_data_normal, gene_data_pathogenic, output_folder)
plot_gene(df, 'DIP2B', gene_data_normal, gene_data_pathogenic, output_folder)
plot_gene(df, 'FMR1_CGG', gene_data_normal, gene_data_pathogenic, output_folder)
plot_gene(df, 'HTT_CAG', gene_data_normal, gene_data_pathogenic, output_folder)
plot_gene(df, 'JPH3_CTG', gene_data_normal, gene_data_pathogenic, output_folder)
plot_gene(df, 'NOP56_GGCCTG', gene_data_normal, gene_data_pathogenic, output_folder)
plot_gene(df, 'PABPN1', gene_data_normal, gene_data_pathogenic, output_folder)
plot_gene(df, 'PPP2R2B_CAG', gene_data_normal, gene_data_pathogenic, output_folder)
plot_gene(df, 'PHOX2B', gene_data_normal, gene_data_pathogenic, output_folder)
plot_gene(df, 'TBP_CAG', gene_data_normal, gene_data_pathogenic, output_folder)


# Printing freq raw numbers
#raw_numbers_output_folder = '/Users/kibanez/Documents/GEL_STR/Pilot/split_by_delivery_version_new_strategy_Feb2018/V1/tables'
#raw_numbers_output_folder = '/Users/kibanez/Documents/GEL_STR/Pilot/split_by_delivery_version_new_strategy_Feb2018/V1andV2/tables'
#raw_numbers_output_folder = '/Users/kibanez/Documents/GEL_STR/Pilot/split_by_delivery_version_new_strategy_Feb2018/V2/tables'
#raw_numbers_output_folder = '/Users/kibanez/Documents/GEL_STR/STR_internal_allele_frequencies/EH-offtarget-v2.5.5-duplicatesRemoved/merged/tables'
#raw_numbers_output_folder = '/Users/kibanez/Documents/GEL_STR/STR_internal_allele_frequencies/august2018/EH-offtarget-v2.5.5-duplicatesRemoved/merged/tables'
raw_numbers_output_folder = '/Users/kibanez/Documents/GEL_STR/STR_internal_allele_frequencies/january2019/EH-offtarget-v2.5.5-duplicatesRemoved/merged/tables'

dir.create(raw_numbers_output_folder)

print_table_gene(df, 'C9orf72_GGGGCC', raw_numbers_output_folder)
print_table_gene(df, 'FMR1_CGG', raw_numbers_output_folder)
print_table_gene(df, 'ATXN2_CAG', raw_numbers_output_folder)
print_table_gene(df, 'ATXN3_CAG', raw_numbers_output_folder)
print_table_gene(df, 'ATN1_CAG', raw_numbers_output_folder)
print_table_gene(df, 'ATXN7_CAG', raw_numbers_output_folder)
print_table_gene(df, 'HTT_CAG', raw_numbers_output_folder)
print_table_gene(df, 'CACNA1A_CAG', raw_numbers_output_folder)
print_table_gene(df, 'PPP2R2B_CAG', raw_numbers_output_folder)
print_table_gene(df, 'ATXN1_CAG', raw_numbers_output_folder)
print_table_gene(df, 'ATXN10_ATTCT', raw_numbers_output_folder)
print_table_gene(df, 'TBP_CAG', raw_numbers_output_folder)
print_table_gene(df, 'TBP2', raw_numbers_output_folder)
print_table_gene(df, 'DMPK', raw_numbers_output_folder)
print_table_gene(df, 'CSTB', raw_numbers_output_folder)
print_table_gene(df, 'AR', raw_numbers_output_folder)
print_table_gene(df, 'ARX', raw_numbers_output_folder)
print_table_gene(df, 'CNBP', raw_numbers_output_folder)
print_table_gene(df, 'FXN1', raw_numbers_output_folder)
print_table_gene(df, 'DIP2B', raw_numbers_output_folder)
print_table_gene(df, 'NOP56', raw_numbers_output_folder)
print_table_gene(df, 'PABPN1', raw_numbers_output_folder)
print_table_gene(df, 'PHOX2B', raw_numbers_output_folder)
print_table_gene(df, 'ATXN8OS', raw_numbers_output_folder)
print_table_gene(df, 'DAB1', raw_numbers_output_folder)
print_table_gene(df, 'JPH3', raw_numbers_output_folder)

