# R script - as input it takes a VCF file after having merged all individual VCF files generated by EH (merging_EH_VCF.py)
# Now merging both b37 and b38 assemblies
date ()
Sys.info ()[c("nodename", "user")]
commandArgs ()
rm (list = ls ())
R.version.string ## "R version 3.6.1 (2019-07-05)"

library(ggplot2); packageDescription ("ggplot2", fields = "Version") #"3.2.1"
library(grid); packageDescription ("grid", fields = "Version") #"3.6.1"
library(gridExtra); packageDescription ("gridExtra", fields = "Version") #"2.3"
library(reshape2); packageDescription ("reshape2", fields = "Version") #"1.4.3"
require(dplyr); packageDescription ("dplyr", fields = "Version") #"0.8.3"

source("~/git/analysing_STRs/functions/plot_gene_mergingAssemblies.R")

# Function that prints into a file the raw data/numbers corresponding to the STR repeat-size frequencies plots
print_table_gene <- function(df_input, gene_name, output_folder){
  df_to_print = df_input %>% filter(gene %in% gene_name)
  df_to_print = df_to_print[,c(4,9)]
  colnames(df_to_print) = c('repeat-size', 'num_samples')
  output_table = paste(output_folder, paste(gene_name, 'freq_rawdata.tsv', sep = '_') , sep = '/')
  write.table(df_to_print, file = output_table, quote = F, row.names = F, col.names = T, sep = '\t')  
}

# Represent in a plot, the allele frequency in Y-coor and the repeat-size in X-coor

# STR annotation, threshold including the largest normal and the smallest pathogenic sizes reported
gene_annotation_normal = '~/git/analysing_STRs/threshold_largest_normal_reported_research.txt'
gene_data_normal = read.table(gene_annotation_normal, stringsAsFactors=F, header = T)

gene_annotation_pathogenic = '~/git/analysing_STRs/threshold_smallest_pathogenic_reported_research.txt'
gene_data_pathogenic = read.table(gene_annotation_pathogenic, stringsAsFactors=F, header = T)

# Research ~80K genomes, EH-v3.1.2- November 2019
setwd("~/Documents/STRs/data/research/batch_march2020/output_EHv2.5.5_vcfs/merged/")
df = read.csv('merged_loci_92665_research_genomes_EHv2.5.5_batch_march2020.tsv',
              sep = '\t',
              stringsAsFactors = F,
              header = T)
dim(df)
# 7857  11

# 1. Merge GRCh37 and GRCh38 info, since chromosome names are different
# GRCh37 are chr1, chr2, chr3 while GRCh38 are 1,2,3
df$chr = recode(df$chr,
                "1" = "chr1",
                "2" = "chr2",
                "3" = "chr3",
                "4" = "chr4",
                "5" = "chr5",
                "6" = "chr6",
                "7" = "chr7",
                "8" = "chr8",
                "9" = "chr9",
                "10" = "chr10",
                "11" = "chr11",
                "12" = "chr12",
                "13" = "chr13",
                "14" = "chr14",
                "15" = "chr15",
                "16" = "chr16",
                "17" = "chr17",
                "18" = "chr18",
                "19" = "chr19",
                "20" = "chr20",
                "21" = "chr21",
                "22" = "chr22",
                "X" = "chrX")
df = df %>%
  group_by(chr, gene, allele) %>%
  mutate(total_num_samples = sum(num_samples)) %>%
  ungroup() %>%
  as.data.frame() 

dim(df)
# 7857  12

df_simpl = df %>% 
  select(chr, gene, allele, total_num_samples)
df_simpl = unique(df_simpl)
dim(df_simpl)
# 5212  4

# Folder where we want to save the plots - PDFs
output_folder = 'plots'
dir.create(output_folder)


# The strategy we were following before was based on the start genomic coordinate, we now use the name of the gene.
l_loci = sort(unique(df_simpl$gene))
for(i in 1:length(l_loci)){
  plot_gene_mergingAssemblies(df_simpl, l_loci[i], output_folder)
}
