# R script - as input it takes a VCF file after having merged all individual VCF files generated by EH (merging_EH_VCF.py)
# Now merging both b37 and b38 assemblies
date ()
Sys.info ()[c("nodename", "user")]
commandArgs ()
rm (list = ls ())
R.version.string ## "R version 3.6.1 (2019-07-05)"

library(ggplot2); packageDescription ("ggplot2", fields = "Version") #"3.2.1"
library(grid); packageDescription ("grid", fields = "Version") #"3.6.1"
library(gridExtra); packageDescription ("gridExtra", fields = "Version") #"2.3"
library(reshape2); packageDescription ("reshape2", fields = "Version") #"1.4.3"
require(dplyr); packageDescription ("dplyr", fields = "Version") #"0.8.3"

source("~/git/analysing_STRs/functions/plot_gene_mergingAssemblies.R")

# Function that prints into a file the raw data/numbers corresponding to the STR repeat-size frequencies plots
print_table_gene <- function(df_input, gene_name, output_folder){
  df_to_print = df_input %>% filter(gene %in% gene_name)
  df_to_print = df_to_print[,c(4,9)]
  colnames(df_to_print) = c('repeat-size', 'num_samples')
  output_table = paste(output_folder, paste(gene_name, 'freq_rawdata.tsv', sep = '_') , sep = '/')
  write.table(df_to_print, file = output_table, quote = F, row.names = F, col.names = T, sep = '\t')  
}

# Represent in a plot, the allele frequency in Y-coor and the repeat-size in X-coor

# STR annotation, threshold including the largest normal and the smallest pathogenic sizes reported
gene_annotation_normal = '~/git/analysing_STRs/threshold_largest_normal_reported_research.txt'
gene_data_normal = read.table(gene_annotation_normal, stringsAsFactors=F, header = T)

gene_annotation_pathogenic = '~/git/analysing_STRs/threshold_smallest_pathogenic_reported_research.txt'
gene_data_pathogenic = read.table(gene_annotation_pathogenic, stringsAsFactors=F, header = T)

# Research ~80K genomes, EH-v3.1.2- November 2019
setwd("~/Documents/STRs/data/research/batch_march2020/output_EHv2.5.5_vcfs/merged/")
df = read.csv('merged_loci_92665_research_genomes_EHv2.5.5_batch_march2020.tsv',
              sep = '\t',
              stringsAsFactors = F,
              header = T)
dim(df)
# 7857  11

# 1. Merge GRCh37 and GRCh38 info, since chromosome names are different
# GRCh37 are chr1, chr2, chr3 while GRCh38 are 1,2,3
df$chr = recode(df$chr,
                "1" = "chr1",
                "2" = "chr2",
                "3" = "chr3",
                "4" = "chr4",
                "5" = "chr5",
                "6" = "chr6",
                "7" = "chr7",
                "8" = "chr8",
                "9" = "chr9",
                "10" = "chr10",
                "11" = "chr11",
                "12" = "chr12",
                "13" = "chr13",
                "14" = "chr14",
                "15" = "chr15",
                "16" = "chr16",
                "17" = "chr17",
                "18" = "chr18",
                "19" = "chr19",
                "20" = "chr20",
                "21" = "chr21",
                "22" = "chr22",
                "X" = "chrX")
df = df %>%
  group_by(chr, gene, allele) %>%
  mutate(total_num_samples = sum(num_samples)) %>%
  ungroup() %>%
  as.data.frame() 

dim(df)
# 7857  12

df_simpl = df %>% 
  select(chr, gene, allele, total_num_samples)
df_simpl = unique(df_simpl)
dim(df_simpl)
# 5212  4

# Folder where we want to save the plots - PDFs
output_folder = 'plots'
dir.create(output_folder)


# This research merged TSV file is special because we do have GRCh37 and GRCh38 genomes altogether
# GRCh37 VCF files have 1,2,3,4...X,Y,MT chromosome nomenclature
# GRCh38 VCF files have chr1, chr2, chr3, ..., chrX, chrX, chrMT nomenclature
# And also, the genomic positions for the loci are different, so we cannot combine them easily

# Let's stratify the whole TSV in b37 and b38
df_b37 = df %>% filter(!grepl("chr", chr))
dim(df_b37)
# 1561 11

df_b38 = df %>% filter(grepl("chr", chr))
dim(df_b38)
# 2422 11
                
# Folder where we want to save the plots - PDFs
output_folder = 'plots'
dir.create(output_folder)

# The strategy we were following before was based on the start genomic coordinate, we now use the name of the gene.
# We don't have genomic coordinates for NOTCH2NLC for GRCh37 (cannot liftover from GRCh38)
l_loci = sort(unique(df$gene))
for(i in 1:length(l_loci)){
  plot_gene(df_b38, l_loci[i], gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")
  plot_gene(df_b37, l_loci[i], gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
}

# Printing freq raw numbers
#raw_numbers_output_folder = '/Users/kibanez/Documents/GEL_STR/Pilot/split_by_delivery_version_new_strategy_Feb2018/V1/tables'
#raw_numbers_output_folder = '/Users/kibanez/Documents/GEL_STR/Pilot/split_by_delivery_version_new_strategy_Feb2018/V1andV2/tables'
#raw_numbers_output_folder = '/Users/kibanez/Documents/GEL_STR/Pilot/split_by_delivery_version_new_strategy_Feb2018/V2/tables'
#raw_numbers_output_folder = '/Users/kibanez/Documents/GEL_STR/STR_internal_allele_frequencies/EH-offtarget-v2.5.5-duplicatesRemoved/merged/tables'
#raw_numbers_output_folder = '/Users/kibanez/Documents/GEL_STR/STR_internal_allele_frequencies/august2018/EH-offtarget-v2.5.5-duplicatesRemoved/merged/tables'
raw_numbers_output_folder = '/Users/kibanez/Documents/GEL_STR/STR_internal_allele_frequencies/january2019/EH-offtarget-v2.5.5-duplicatesRemoved/merged/tables'

dir.create(raw_numbers_output_folder)

print_table_gene(df, 'C9orf72_GGGGCC', raw_numbers_output_folder)
print_table_gene(df, 'FMR1_CGG', raw_numbers_output_folder)
print_table_gene(df, 'ATXN2_CAG', raw_numbers_output_folder)
print_table_gene(df, 'ATXN3_CAG', raw_numbers_output_folder)
print_table_gene(df, 'ATN1_CAG', raw_numbers_output_folder)
print_table_gene(df, 'ATXN7_CAG', raw_numbers_output_folder)
print_table_gene(df, 'HTT_CAG', raw_numbers_output_folder)
print_table_gene(df, 'CACNA1A_CAG', raw_numbers_output_folder)
print_table_gene(df, 'PPP2R2B_CAG', raw_numbers_output_folder)
print_table_gene(df, 'ATXN1_CAG', raw_numbers_output_folder)
print_table_gene(df, 'ATXN10_ATTCT', raw_numbers_output_folder)
print_table_gene(df, 'TBP_CAG', raw_numbers_output_folder)
print_table_gene(df, 'TBP2', raw_numbers_output_folder)
print_table_gene(df, 'DMPK', raw_numbers_output_folder)
print_table_gene(df, 'CSTB', raw_numbers_output_folder)
print_table_gene(df, 'AR', raw_numbers_output_folder)
print_table_gene(df, 'ARX', raw_numbers_output_folder)
print_table_gene(df, 'CNBP', raw_numbers_output_folder)
print_table_gene(df, 'FXN1', raw_numbers_output_folder)
print_table_gene(df, 'DIP2B', raw_numbers_output_folder)
print_table_gene(df, 'NOP56', raw_numbers_output_folder)
print_table_gene(df, 'PABPN1', raw_numbers_output_folder)
print_table_gene(df, 'PHOX2B', raw_numbers_output_folder)
print_table_gene(df, 'ATXN8OS', raw_numbers_output_folder)
print_table_gene(df, 'DAB1', raw_numbers_output_folder)
print_table_gene(df, 'JPH3', raw_numbers_output_folder)

