# R script - as input it takes a VCF file after having merged all individual VCF files generated by EH (merging_EH_VCF.py)
date ()
Sys.info ()[c("nodename", "user")]
commandArgs ()
rm (list = ls ())
R.version.string ## "R version 3.6.1 (2019-07-05)"

library(ggplot2); packageDescription ("ggplot2", fields = "Version") #"3.2.1"
library(grid); packageDescription ("grid", fields = "Version") #"3.6.1"
library(gridExtra); packageDescription ("gridExtra", fields = "Version") #"2.3"
library(reshape2); packageDescription ("reshape2", fields = "Version") #"1.4.3"
require(dplyr); packageDescription ("dplyr", fields = "Version") #"0.8.3"

source("~/git/analysing_STRs/functions/plot_gene.R")

# Function that prints into a file the raw data/numbers corresponding to the STR repeat-size frequencies plots
print_table_gene <- function(df_input, gene_name, output_folder){
  df_to_print = df_input %>% filter(gene %in% gene_name)
  df_to_print = df_to_print[,c(4,9)]
  colnames(df_to_print) = c('repeat-size', 'num_samples')
  output_table = paste(output_folder, paste(gene_name, 'freq_rawdata.tsv', sep = '_') , sep = '/')
  write.table(df_to_print, file = output_table, quote = F, row.names = F, col.names = T, sep = '\t')  
}

# Represent in a plot, the allele frequency in Y-coor and the repeat-size in X-coor

# STR annotation, threshold including the largest normal and the smallest pathogenic sizes reported
gene_annotation_normal = './threshold_largest_normal_reported_research.txt'
gene_data_normal = read.table(gene_annotation_normal, stringsAsFactors=F, header = T)

gene_annotation_pathogenic = './threshold_smallest_pathogenic_reported_research.txt'
gene_data_pathogenic = read.table(gene_annotation_pathogenic, stringsAsFactors=F, header = T)

# Research ~80K genomes, EH-v3.1.2- November 2019
setwd("~/Documents/STRs/data/research/EH_3.1.2_research_October2019/merged_genotypeUpdated//")
df = read.csv('./merged_loci_82565_research_genomes_EHv3.1.2_october2019.tsv',
              sep = '\t',
              stringsAsFactors = F,
              header = T)
dim(df)
# 4495 12


# This research merged TSV file is special because we do have GRCh37 and GRCh38 genomes altogether
# GRCh37 VCF files have 1,2,3,4...X,Y,MT chromosome nomenclature
# GRCh38 VCF files have chr1, chr2, chr3, ..., chrX, chrX, chrMT nomenclature
# And also, the genomic positions for the loci are different, so we cannot combine them easily

# Let's stratify the whole TSV in b37 and b38
df_b37 = df %>% filter(!grepl("chr", chr))
dim(df_b37)
# 1868 12

df_b38 = df %>% filter(grepl("chr", chr))
dim(df_b38)
# 2627 12
                
# Folder where we want to save the plots - PDFs
output_folder = 'plots'
dir.create(output_folder)

# The strategy we were following before was based on the start genomic coordinate, we now use the name of the gene.
# We don't have genomic coordinates for NOTCH2NLC for GRCh37 (cannot liftover from GRCh38)
plot_gene(df_b37, 'AFF2', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'AFF2', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b38, 'AFF2_GEL', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'AR', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'AR', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'AR_GEL', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'AR_GEL', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'ATN1', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'ATN1', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'ATXN1', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'ATXN1', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'ATXN2', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'ATXN2', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'ATXN3', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'ATXN3', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'ATXN7', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'ATXN7', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'ATXN7_GCC', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'ATXN7_GCC', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'ATXN8OS', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'ATXN8OS', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'ATXN8OS_CTA', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'ATXN8OS_CTA', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'ATXN10', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'ATXN10', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'C11ORF80', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'C11ORF80', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'C9ORF72', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'C9ORF72', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'CACNA1A', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'CACNA1A', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'CBL', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'CBL', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'CNBP', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'CNBP', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'CNBP_CAGA', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'CNBP_CAGA', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'CRACR2A', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'CRACR2A', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'CSTB', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'CSTB', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'DIP2B', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'DIP2B', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'DMPK', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'DMPK', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'FAME1', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'FAME1', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'FAME2', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'FAME2', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'FAME3', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'FAME3', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")


plot_gene(df_b37, 'FMR1', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'FMR1', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'FRA10AC1', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'FRA10AC1', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'FXN', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'FXN', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'GLS', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'GLS', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'HTT', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'HTT', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'HTT_CCG', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'HTT_CCG', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'JPH3', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'JPH3', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'LINGO3', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'LINGO3', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'MARCH6', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'MARCH6', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'NOP56', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'NOP56', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'NOP56_CGCCTG', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'NOP56_CGCCTG', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'PCMTD2', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'PCMTD2', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'PHOX2B', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'PHOX2B', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'PPP2R2B', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'PPP2R2B', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'RFC1', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'RFC1', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'TBP', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'TBP', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

plot_gene(df_b37, 'TCF4', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh37", "")
plot_gene(df_b38, 'TCF4', gene_data_normal, gene_data_pathogenic, output_folder, "GRCh38", "")

# Printing freq raw numbers
#raw_numbers_output_folder = '/Users/kibanez/Documents/GEL_STR/Pilot/split_by_delivery_version_new_strategy_Feb2018/V1/tables'
#raw_numbers_output_folder = '/Users/kibanez/Documents/GEL_STR/Pilot/split_by_delivery_version_new_strategy_Feb2018/V1andV2/tables'
#raw_numbers_output_folder = '/Users/kibanez/Documents/GEL_STR/Pilot/split_by_delivery_version_new_strategy_Feb2018/V2/tables'
#raw_numbers_output_folder = '/Users/kibanez/Documents/GEL_STR/STR_internal_allele_frequencies/EH-offtarget-v2.5.5-duplicatesRemoved/merged/tables'
#raw_numbers_output_folder = '/Users/kibanez/Documents/GEL_STR/STR_internal_allele_frequencies/august2018/EH-offtarget-v2.5.5-duplicatesRemoved/merged/tables'
raw_numbers_output_folder = '/Users/kibanez/Documents/GEL_STR/STR_internal_allele_frequencies/january2019/EH-offtarget-v2.5.5-duplicatesRemoved/merged/tables'

dir.create(raw_numbers_output_folder)

print_table_gene(df, 'C9orf72_GGGGCC', raw_numbers_output_folder)
print_table_gene(df, 'FMR1_CGG', raw_numbers_output_folder)
print_table_gene(df, 'ATXN2_CAG', raw_numbers_output_folder)
print_table_gene(df, 'ATXN3_CAG', raw_numbers_output_folder)
print_table_gene(df, 'ATN1_CAG', raw_numbers_output_folder)
print_table_gene(df, 'ATXN7_CAG', raw_numbers_output_folder)
print_table_gene(df, 'HTT_CAG', raw_numbers_output_folder)
print_table_gene(df, 'CACNA1A_CAG', raw_numbers_output_folder)
print_table_gene(df, 'PPP2R2B_CAG', raw_numbers_output_folder)
print_table_gene(df, 'ATXN1_CAG', raw_numbers_output_folder)
print_table_gene(df, 'ATXN10_ATTCT', raw_numbers_output_folder)
print_table_gene(df, 'TBP_CAG', raw_numbers_output_folder)
print_table_gene(df, 'TBP2', raw_numbers_output_folder)
print_table_gene(df, 'DMPK', raw_numbers_output_folder)
print_table_gene(df, 'CSTB', raw_numbers_output_folder)
print_table_gene(df, 'AR', raw_numbers_output_folder)
print_table_gene(df, 'ARX', raw_numbers_output_folder)
print_table_gene(df, 'CNBP', raw_numbers_output_folder)
print_table_gene(df, 'FXN1', raw_numbers_output_folder)
print_table_gene(df, 'DIP2B', raw_numbers_output_folder)
print_table_gene(df, 'NOP56', raw_numbers_output_folder)
print_table_gene(df, 'PABPN1', raw_numbers_output_folder)
print_table_gene(df, 'PHOX2B', raw_numbers_output_folder)
print_table_gene(df, 'ATXN8OS', raw_numbers_output_folder)
print_table_gene(df, 'DAB1', raw_numbers_output_folder)
print_table_gene(df, 'JPH3', raw_numbers_output_folder)

